sway-core/src/declaration_engine/declaration_engine.rs

```diff
@@ -158,26 +211,69 @@ impl DeclarationEngine {
         self.slab.get(*index).expect_struct(span)
     }
 
-    fn de_add_monomorphized_struct_copy(
+    pub(crate) fn de_get_monomorphized_struct(
         &self,
         original_id: DeclarationId,
-        new_copy: TypedStructDeclaration,
-    ) {
-        let span = new_copy.span();
-        let new_id =
-            DeclarationId::new(self.slab.insert(DeclarationWrapper::Struct(new_copy)), span);
-        self.de_add_monomorphized_copy(original_id, new_id)
+        type_arguments: &Vec<TypeArgument>,
+        span: &Span,
+    ) -> Option<TypedStructDeclaration> {
+        for copy in self
+            .de_get_monomorphized_copies(original_id)
+            .iter()
+            .cloned()
+        {
+            let monomorphized_struct_decl = copy.expect_struct(span).unwrap();
+            if &monomorphized_struct_decl.type_parameters == type_arguments {
```

Comment:
Same as above---most generic structs are called without type arguments, so most of the time this check will fail, missing valid cache opportunities.